File: CPP/572. SubtreeofAnotherTree.cpp
#include <bits/stdc++.h>
using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

void preOrderVisit(TreeNode* root) {
    if (root==nullptr) return;
    cout << root->val << " ";
    preOrderVisit(root->left);
    preOrderVisit(root->right);
}


bool isSameTree(TreeNode* p, TreeNode* q) {
    // cout << "Testing" << endl;
    // preOrderVisit(p);
    // cout << endl;
    // preOrderVisit(q);
    // cout << endl;

    if (p==nullptr && q==nullptr) return true;
    else if (p==nullptr || q==nullptr) return false;
    else if (p->val != q->val) return false;
    else {
        bool left=isSameTree(p->left,q->left);
        bool right=isSameTree(p->right,q->right);
        return left && right;
    }
}

bool isSubtree(TreeNode* root, TreeNode* subRoot) {
    if (root==nullptr) return false;
    if (root->val==subRoot->val) {
        if (isSameTree(root,subRoot)) return true;
    }
    return isSubtree(root->left,subRoot) || isSubtree(root->right,subRoot);
}

int main() {
    TreeNode *bt1 = new TreeNode(3,new TreeNode(4,new TreeNode(1),new TreeNode(2)),new TreeNode(5));
    TreeNode *bt2 = new TreeNode(4,new TreeNode(1),new TreeNode(2));
    cout << isSubtree(bt1,bt2) << endl; // true

    TreeNode *bt3 = new TreeNode(3,new TreeNode(4,new TreeNode(1),new TreeNode(2,new TreeNode(0),nullptr)),new TreeNode(5));
    TreeNode *bt4 = new TreeNode(4,new TreeNode(1),new TreeNode(2));
    cout << isSubtree(bt3,bt4) << endl; // false

    TreeNode *bt5 = new TreeNode(1,new TreeNode(1),nullptr);
    TreeNode *bt6 = new TreeNode(1);
    cout << isSubtree(bt5,bt6) << endl; // true
    
    return 0;
}

File: Python/572.SubtreeofAnotherTree.py
from typing import Optional
from collections import deque

def printTreePre(t):
    if not t: return
    print(t.val,sep=' ')
    printTreePre(t.left)
    printTreePre(t.right)

def array_to_tree(arr):
    if not arr:
        return None

    root = TreeNode(arr[0])
    queue = deque([root])
    i = 1

    while queue and i < len(arr):
        node = queue.popleft()

        if arr[i] is not None:
            node.left = TreeNode(arr[i])
            queue.append(node.left)
        i += 1

        if i < len(arr) and arr[i] is not None:
            node.right = TreeNode(arr[i])
            queue.append(node.right)
        i += 1

    return root

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q: return True
        if p and q:
            if p.val!=q.val: return False
            return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
        else:
            return False
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root or not subRoot: return False
        if root.val==subRoot.val:
            if self.isSameTree(root,subRoot): return True
        if root:
            return self.isSubtree(root.left,subRoot) or self.isSubtree(root.right,subRoot)
        return False

s=Solution()
print(s.isSubtree(array_to_tree([3,4,5,1,2]),array_to_tree([4,1,2]))) # True
print(s.isSubtree(array_to_tree([3,4,5,1,2,None,None,None,None,0]),array_to_tree([4,1,2]))) # False


